### DISCUSSION TITLE
On circular imports, use of forward references / `model_rebuild` in schemas

### DISCUSSION THREAD

**Discussion by zzstoatzz** [PrefectHQ/prefect#16387] [General]
This is a place to discuss and form a plan to improve:
- issues with circular imports (we defer too much as a result of these issues)
- spread use of `model_rebuild()` it should only happen in `prefect.main` ideally

related issues / PRs:
- https://github.com/PrefectHQ/prefect/issues/15957
- https://github.com/PrefectHQ/prefect/issues/15047
- https://github.com/PrefectHQ/prefect/pull/16089
- https://github.com/PrefectHQ/prefect/pull/16434
- https://github.com/PrefectHQ/prefect/issues/18317

**Comment by aaazzam** [comment]
you want to build a ... DAG ... of dependencies?

**Reply by aaazzam** [reply]
Also #16434

**Reply by aaazzam** [reply]
Also #15957

**Reply by zzstoatzz** [reply]
👍 #15957 is the first one linked in the body - ill move #16434 in there too

**Comment by zzstoatzz** [comment]
https://github.com/PrefectHQ/prefect/pull/19344

**Comment by joshuastagner** [comment]
Hit a case of this adding cancellation logic to the `prefect-docker` integration. I added a new pathway that will have the worker cancel a flow-run. If a worker attempts to cancel a flow run without first submitting a new flow run, it throw the `TypeError: 'MockValSer' object cannot be converted to 'SchemaSerializer'` error. By rebuilding the nested `StateDetails` fixes this error. The code in question was never merged, but hit this while testing a new change locally
